### ISSUE TITLE
Using apiserver --goaway-chance reduces apiservice request throughput by 8%

### ISSUE THREAD

**Issue by nrockwell3** [kubernetes/kubernetes#90821]
**What happened**:
Using the apiserver `--goaway-chance=0.001` flag resulted in our apiserver load tests reported a 8% reduction in throughput vs a baseline without the flag. One thought was that the GOAWAY operations on a connections were the source of the reduction. So we ran the test with `--goaway-chance=0.0001`. This resulted in the same 8% reduction in throughput. This leads to the conclusion that just enabling `--goaway-chance` resulted in the 8% reduction in throughput.

Test results from an hour long jmeter run:
<img width="1096" alt="Screen Shot 2020-05-06 at 10 29 02 AM" src="https://user-images.githubusercontent.com/22150226/81211155-f9666c80-8fa0-11ea-917b-cfc1efffc05f.png">

**What you expected to happen**:
We expect that the overhead for using `--goaway-chance` would be minimal (i.e. <1% reduction in throughput).

**How to reproduce it (as minimally and precisely as possible)**:
The test uses 50 jmeter servers to drive the load. Each server does HTTPS GET requests (/api/v1/namespaces and /api/v1/pods). The requests hit 10 different kube 1.18 deployments, each of which has 3 apiservers for HA. I'd expect any test that drives the apiserver hard will show the same results.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.18.2+IKS
- Cloud provider or hardware configuration: IBM Cloud Kubernetes Service
- OS (e.g: `cat /etc/os-release`): Debian GNU/Linux 10 (buster)
- Kernel (e.g. `uname -a`): 4.15.0-96-generic #97-Ubuntu SMP Wed Apr 1 03:25:46 UTC 2020 x86_64 GNU/Linux
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by liggitt** [comment]
/sig api-machinery
cc @deads2k @lavalamp @smarterclayton

**Comment by lavalamp** [comment]
What are the units in the table? What column is the std devation of?

What are the double entries for, did you run each scenario twice?

How many requests were done in each run? What was the duration of each run?

The result is weird enough that I suspect a measuring artifact. Can you try .000000001?

**Comment by nrockwell3** [comment]
@lavalamp
Units are milliseconds. Standard deviation of response times.

Double entries are a result of running each scenario twice. The "Throughput as % baseline" calculations are based on the average of the throughput for the two samples.

The number of # samples indicates the number of requests in each run. The tests run for an hour. 

I will run again with .000000001

**Comment by wojtek-t** [comment]
> The result is weird enough that I suspect a measuring artifact. Can you try .000000001?

+1 - changing the value should result in different numbers

**Comment by liggitt** [comment]
unless the dominating factor is a constant-cost performance penalty from use of the goaway chance sync.Pool

**Comment by nrockwell3** [comment]
I ran the experiment with `--goaway-chance=0.000000001` and got the best throughput (blue line in table) of any test. 
<img width="963" alt="Screen Shot 2020-05-11 at 12 12 02 PM" src="https://user-images.githubusercontent.com/22150226/81585252-e2998e80-9381-11ea-9cf2-b851925dafe2.png">

**Comment by lavalamp** [comment]
I strongly suspect the difference is down to luck about the particulars of where your worker/control plane nodes ended up.

I say that because I doubt you made anything close to a billion requests during this test and therefore not even one GOAWAY should have been sent. (It's why I picked the number 1/billion.) Therefore you should be paying the cost of the sync.Pool or whatever else, but not reaping any benefit at all. Yet, the latency went down, which makes no sense. If the tests were consistent, we should have seen numbers between the "off" and .001 cases.

Basically, there should be an effect here (or we wouldn't have added the feature!) but your tests are not powerful enough to identify its sign.

I suggest running more, smaller tests, at different times of day, on different VMs, in different data centers etc. With this small number of tests, I think you're actually measuring your environment and not the flags. 8% variability--or even twice that--from the environment is totally believable.

(also sync.Pool does per-thread magic and is super efficient--basically no locks used.)

**Comment by nrockwell3** [comment]
I've run additional tests and agree that what I'm seeing is within the normal variation dependent on placement, ect. Thanks for help with this.
